### PR TITLE
fix GEARPUMP-122, refactor kafka connector API

### PR DESCRIPTION
**Pull request by manuzhang**
This PR mainly refactor the Kafka connector API with the following changes:

1. removes `OffsetStorage` and `KafkaStorage`, and uses `CheckpointStore` and `KafkaStore` for both offset and state checkpoint stores.
2. adds a `checkpoint(CheckpointStoreFactory)` method to `TimeReplayableSource` which makes offset store configurable by user. If `checkpoint` is not called, then no offset checkpoint will be performed and applications run in *at-most-once* mode.
3. moves user facing APIs,  `KafkaSource`, `KafkaSink` and `KafkaStore` to Java and keeps their implementations in Scala, which prevents Scala package private methods being exposed in Java programs. 
4.  `KafkaSource`, `KafkaSink` and `KafkaStore` are all configurable through Java properties which is the standard Kafka way.
5. update corresponding UTs.

### CODE REVIEW

File: external/kafka/src/main/scala/org/apache/gearpump/streaming/kafka/lib/source/AbstractKafkaSource.scala [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+    Map.empty[TopicAndPartition, CheckpointStore]
+
+  override def checkpoint(checkpointStoreFactory: CheckpointStoreFactory): Unit = {
+    this.checkpointStoreFactory = Some(checkpointStoreFactory)
+  }
+
+  override def open(context: TaskContext, startTime: TimeStamp): Unit = {
+    import context.{parallelism, taskId}
+
+    LOG.info("KafkaSource opened at start time {}", startTime)
```
Comment (huafengw) [inline external/kafka/src/main/scala/org/apache/gearpump/streaming/kafka/lib/source/AbstractKafkaSource.scala]: I think adding the topic list that are consumed to the log is better.

Comment (manuzhang) [inline external/kafka/src/main/scala/org/apache/gearpump/streaming/kafka/lib/source/AbstractKafkaSource.scala]: We already log assigned partitions below. Will that be sufficient ?

Comment (huafengw) [inline external/kafka/src/main/scala/org/apache/gearpump/streaming/kafka/lib/source/AbstractKafkaSource.scala]: all right, I missed that part.

**Comment by huafengw** [comment]
Documentation update?

**Comment by manuzhang** [comment]
may I update doc as follow-up ?

File: examples/streaming/kafka/src/main/scala/org/apache/gearpump/streaming/examples/kafka/KafkaReadWrite.scala [hunk 2]
```diff
@@ -60,14 +64,19 @@ object KafkaReadWrite extends AkkaApp with ArgumentsParser {
...
     val appConfig = UserConfig.empty
-    val offsetStorageFactory = new KafkaStorageFactory(zookeeperConnect, brokerList)
-    val source = new KafkaSource(sourceTopic, zookeeperConnect, offsetStorageFactory)
+    val props = new Properties
+    props.put(KafkaConfig.ZOOKEEPER_CONNECT_CONFIG, zookeeperConnect)
+    props.put(KafkaConfig.BOOTSTRAP_SERVERS_CONFIG, brokerList)
+    props.put(KafkaConfig.CHECKPOINT_STORE_NAME_PREFIX_CONFIG, appName)
+    val source = new KafkaSource(sourceTopic, props)
+    val checkpointStoreFactory = new KafkaStoreFactory(props)
+    source.checkpoint(checkpointStoreFactory)
```
Comment (huafengw) [inline examples/streaming/kafka/src/main/scala/org/apache/gearpump/streaming/examples/kafka/KafkaReadWrite.scala]: What if user doesn't call ```checkpoint```?

Comment (manuzhang) [inline examples/streaming/kafka/src/main/scala/org/apache/gearpump/streaming/examples/kafka/KafkaReadWrite.scala]: then kafka offsets will not be checkpointed

File: external/kafka/src/main/java/org/apache/gearpump/streaming/kafka/KafkaSource.java [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+      KafkaConfig.KafkaConfigFactory configFactory,
+      KafkaClient.KafkaClientFactory clientFactory,
+      FetchThread.FetchThreadFactory threadFactory) {
+    super(topic, properties, configFactory, clientFactory, threadFactory);
+  }
+
+  /**
+   * for tests only
+   */
+  void addPartitionAndStore(TopicAndPartition tp, CheckpointStore store) {
```
Comment (huafengw) [inline external/kafka/src/main/java/org/apache/gearpump/streaming/kafka/KafkaSource.java]: ```protected```?

**Comment by manuzhang** [comment]
please do not merge. I need to verify whether integration test passes

**Comment by manuzhang** [comment]
I've verified related integration tests pass

**Comment by huafengw** [comment]
UT failed

**Comment by huafengw** [comment]
+1

**Pull request closed without merging**